src/main/cpp/win_fsnotifier.cpp

```diff
@@ -51,99 +176,96 @@ void handlePathChanged(watch_details_t *details, FILE_NOTIFY_INFORMATION *info)
     } else if (info->Action == FILE_ACTION_MODIFIED) {
         type = FILE_EVENT_MODIFIED;
     } else {
-        wprintf(L"~~~~ Unknown event 0x%x for %ls\n", info->Action, changedPath);
+        wprintf(L"~~~~ Unknown event 0x%x for %ls\n", info->Action, changedPath.c_str());
         type = FILE_EVENT_UNKNOWN;
     }
 
-    bool watching = false;
-    for (int i = 0; i < details->watchedPathCount; i++) {
-        wchar_t* watchedPath = details->watchedPaths[i];
-        if (wcsncmp(watchedPath, changedPath, wcslen(watchedPath)) == 0) {
-            watching = true;
-            break;
-        }
-    }
-    if (!watching) {
-        wprintf(L"~~~~ Ignoring %ls (root is not watched)\n", changedPath);
-        return;
+    server->reportEvent(type, changedPath);
+}
+
+void Server::startWatching(wchar_t *path) {
+    WatchPoint* watchPoint = new WatchPoint(this, wstring(path));
+    QueueUserAPC(&startWatchCallback, threadHandle, (ULONG_PTR) watchPoint);
+}
+
+// Called by QueueUserAPC to add another directory.
+static void CALLBACK startWatchCallback(_In_ ULONG_PTR arg) {
+    WatchPoint* watchPoint = (WatchPoint*)arg;
+    watchPoint->server->watchPoints.push_back(watchPoint);
+    watchPoint->listen();
+}
+
+void Server::reportFinished(WatchPoint* watchPoint) {
+    watchPoints.remove(watchPoint);
+}
+
+static JNIEnv* getThreadEnv(JavaVM *jvm) {
+    JNIEnv* env;
+    // TODO Verify that JNI 1.6 is the right version
+    jint ret = jvm->GetEnv((void **) &(env), JNI_VERSION_1_6);
+    if (ret != JNI_OK) {
+        printf("Failed to attach JNI to current thread: %d\n", ret);
+        return NULL;
     }
+    return env;
+}
 
-    reportEvent(type, changedPath, changedPathLen, details);
-    free(changedPath);
+void Server::reportEvent(jint type, const wstring changedPath) {
+    JNIEnv* env = getThreadEnv(jvm);
+    jstring changedPathJava = wchar_to_java_path(env, changedPath.c_str());
```

Comment:
❓ It is unclear if the `jstring` should be cleanup here or not. According to the documentation, a Java string is cleared up upon returning from the native call. However, in this case, the call back is originating from native. You may have to do some clean up here.

Reply:
I don't know what I'd need to clean up here. Can you point me the documentation you are referring to?

Reply:
In the `wchar_to_java_path` method, there is call to `env->NewString` (https://github.com/adammurdoch/native-platform/pull/62/files#diff-6d6dcecdbdda9917656c3e4edbaa0c02R66). The Oracle JNI documentation is really poor on the contract of the `NewString` method (https://docs.oracle.com/en/java/javase/13/docs/specs/jni/functions.html#newstring). If I read this correctly, the object returned will have some allocated resources which will now be owned by the `reportEvent` method. Assuming `jstring` works the same way as a `String` object, the method called by `CallVoidMethod` would only receive a reference of the object. If it saves a reference of the object, only the reference count would increase. Regardless, once the method finish executing the there is still one reference held by `changedPathJava`. I found this comment for the Android NDK (https://groups.google.com/d/msg/android-ndk/hxe3hFvMV1g/_lvsAetcJ2cJ) which I assume would apply to the JDK. According to the comment, the reference is cleanup only when the JNI call returned or an explicit call to `DeleteLocalRef`. In this case, the call originate from a native thread which never return from a JNI call. This would lead in theory to a reference leak.

All that being said, we could double check to see if the jstring object automatically clean it's reference once it goes out of scope inside it's destructor. If it does, then everything should be fine.

Either way, the JNI function documentation is really bad and doesn't explain the ownership which is a problem with the entire JNI documentation.

Reply:
I think this explains well, and IIUC there's no need to do anything: https://stackoverflow.com/a/24292867/104894